pkg/k8s/watchers/subscriber/ciliumnode.go

```diff
@@ -28,19 +34,49 @@ type CiliumNode interface {
 type CiliumNodeChain struct {
 	list
 
+	res  k8s.LocalCiliumNodeResource
 	subs []CiliumNode
 }
 
-// NewCiliumNodeChain creates a CiliumNodeChain ready for its
+// newCiliumNodeChain creates a CiliumNodeChain ready for its
 // Register method to be called.
-func NewCiliumNodeChain() *CiliumNodeChain {
-	return &CiliumNodeChain{}
+func newCiliumNodeChain(res k8s.LocalCiliumNodeResource) *CiliumNodeChain {
+	return &CiliumNodeChain{res: res}
+}
+
+func (l *CiliumNodeChain) process() {
+	events := l.res.Events(context.Background())
+
+	var existing *ciliumv2.CiliumNode
+	for ev := range events {
+		var err error
+
+		switch ev.Kind {
+		case resource.Upsert:
+			node := ev.Object
+			if existing == nil {
+				err = l.OnAddCiliumNode(node)
+			} else {
+				err = l.OnUpdateCiliumNode(existing, node)
+			}
+			existing = node
+		case resource.Delete:
+			err = l.OnDeleteCiliumNode(ev.Object)
+		}
+
+		ev.Done(err)
```

Comment:
This will now retry the event for all subscribers even if only one of them failed. I'm not sure we can rely on them being idempotent. It might be better to keep the original semantics of not doing much with the error (it was just marking up the EventProcessed metric). Though then we won't have `Resource[T]` marking up the metric and will lose metric coverage :(

Take a look at the subscribers and see what options we have.

Reply:
Thanks for this catch. Rewrote the PR to drop CiliumNodeChain completely